docs/proposals/20240930-machine-drain-rules.md

```diff
@@ -0,0 +1,343 @@
+care if they're gracefully terminated and they should run until the Node goes offline (e.g. monitoring agents).
+
+**Story 4: Define drain order & exclude Pods from drain without management cluster access**
+
+As a workload cluster admin/user, I want to be able to define drain order of Pods and also exclude Pods from drain 
+(similar to Story 1-3, just without management cluster access).
+
+**Story 5: Apply drain configuration to a set of Machines / Clusters**
+
+As a cluster admin/user, I would like to be able to configure the drain behavior for all Machines of a management 
+cluster or for a subset of these Machines without having to configure each Machine individually. I would also like 
+to be able to change the drain configuration without having to modify all Machine objects.
+
+### Goals
+
+* Allow users to customize how/if Pods are evicted:
+  * Allow to define the order in which Pods are evicted
+  * Allow to define if Pods should be excluded from drain
+* Use the current drain default rules as a fallback if neither `MachineDrainRule` CRs nor drain annotations
+  apply to a Pod (this also ensures compatibility with current behavior)
+* Stop waiting for volume detachment for Pods that are excluded from drain
+
+### Non-Goals
+
+* Change the drain behavior for DaemonSet and static Pods (we’ll continue to skip draining for both).
+  While the drain behavior itself won't be changed, we will stop waiting for detachment of volumes of DaemonSet Pods.
+
+### Future work
+
+* Align with the evolving landscape for Node drain in Kubernetes as soon as the new features are available, 
+  stable and can be used for the Kubernetes versions Cluster API supports (e.g. [KEP-4212](https://github.com/kubernetes/enhancements/issues/4212), [KEP-4563](https://github.com/kubernetes/enhancements/issues/4563)).
+
+  Because of the wide range of workload cluster Kubernetes versions supported in Cluster API (n-5), it will take 
+  significant time until we can use new Kubernetes features in all supported versions. So an implementation in
+  Cluster API is the only option to improve drain anytime soon (even for the oldest supported Kubernetes versions).
+
+  Also, please note that while we would like to align with Kubernetes and leverage its new features as soon as 
+  possible, there will always be a delta to be addressed in Cluster API due to the fact that Cluster API is operated 
+  from the management cluster, while drain is done in workload clusters.
+
+  Additionally, Cluster API should take care of many clusters, while Kubernetes features are by design scoped to a single cluster.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+#### `MachineDrainRule` CRD
+
+We propose to introduce a new `MachineDrainRule` namespace-scoped CRD. The goal of this CRD is to allow management 
+cluster admins/users to define a drain rule (a specific drain behavior) for a set of Pods.
```

Comment:
How does the controller know which cluster to apply the MDR too? Is it any and all clusters in a namespace? Or is it expected that the machine selector should be used to filter down to machines within clusters you care about somehow?

Reply:
If the machine selector is not set, it will apply to all Machines in the namespace.

If the machine selector is set it will only apply to the Machines matching the selector.

The idea is that the Machine controller lists all MDRs and checks which of these are matching the current Machine.

Reply:
Re-read your comment. There is also a clusterSelector that can be used

Reply:
Ahh yes, missed that! 🤦